### ISSUE TITLE
flying vertices question

### ISSUE THREAD

**Issue by garylidd** [noahcao/Pixel2Mesh#1]
Hi,
Thanks for your sharing, I've been doing the same thing recently, but I got some strange results in my project.The mesh model looks well in original view, like this:
![snapshot00](https://user-images.githubusercontent.com/48756786/62768672-f13cd780-bac9-11e9-8cbb-baee8b0fa6c8.png)
But in other view, I found several flying vertices, like this: 
![snapshot01](https://user-images.githubusercontent.com/48756786/62768676-f568f500-bac9-11e9-9ab2-4d18b19792c5.png)
I've compared my code with yours, the main difference is that I use pytorch_geometric library for GCN, and I use the same number of resblock in official Pixel2Mesh project.
Have you ever encountered this problem ? or any suggestions ? Thanks

**Comment by ultmaster** [comment]
Looks like something wrong with regularization. Did you encounter this problem during training or validation? Would you mind sharing the snippet of code where you define your loss?

**Comment by garylidd** [comment]
I only save mesh during training. Here's my loss function, loss weights are identical with official P2M. Chamfer loss from https://github.com/chrdiller/pyTorchChamferDistance. Thanks ~
```
def laplace_coord(coord, lap_idx):
    vertex = torch.cat([coord, torch.zeros([1,3]).type_as(coord)], 0)
    indices = lap_idx[:, :8]
    weights = lap_idx[:, -1].float()
    weights = torch.reciprocal(weights).view([-1, 1]).repeat(1, 3)
    laplace = torch.sum(vertex[indices], 1)
    laplace = coord - torch.mul(laplace, weights)
    return laplace

def compute_laplace_loss(pre_coord, coord, lap_idx, is_first):
    lap1 = laplace_coord(pre_coord, lap_idx)
    lap2 = laplace_coord(coord, lap_idx)
    laplace_loss = torch.mean(torch.sum((lap1 - lap2) ** 2, 1)) * 1500

    if is_first:
        move_loss = torch.Tensor([0]).type_as(laplace_loss)
    else:
        move_loss = torch.mean(torch.sum((pre_coord - coord) ** 2, 1)) * 100
    
    return laplace_loss, move_loss

def compute_normal_loss(normal, min_idx, edge, edge_vec):
    normal = normal.squeeze(0)
    min_idx = min_idx.squeeze(0)
    edge = edge.transpose(0, 1)
    normal_vec = normal[min_idx][edge[:, 0]]
    normalize = F.normalize
    cos_normal = torch.mul(normalize(normal_vec, dim=1), normalize(edge_vec, dim=1))
    cos_normal_loss = torch.mean(torch.abs(torch.sum(cos_normal, 1))) * 0.5
    return cos_normal_loss

def compute_edge_loss(coord, edge):
    edge = edge.transpose(0, 1)
    coord0 = coord[edge[:, 0]]
    coord1 = coord[edge[:, 1]]
    edge_vec = coord0 - coord1
    return edge_vec, torch.mean(torch.sum(edge_vec ** 2, 1))

def compute_chamfer_loss(gt_point, coord):
    chamfer_dist = ChamferDistance()
    return chamfer_dist(gt_point, coord)

def compute_loss(coords, gt_mesh, init_mesh):

    total_loss = 0
    loss_dict = {}
    for i in range(3):
        # edge loss (penalize long edge)
        edge_vec, edge_loss = compute_edge_loss(coords[i], init_mesh.edges[i])
        
        # chamfer Loss (attention to the input order)
        chamfer_loss = compute_chamfer_loss(gt_mesh.pos.unsqueeze(0), coords[i].unsqueeze(0))
        dist1, dist2, idx1, idx2 = chamfer_loss
        point_loss = (torch.mean(dist1) + 0.55 * torch.mean(dist2)) * 3000

        # normal loss 
        norm_loss = compute_normal_loss(gt_mesh.norm, idx2.long(), init_mesh.edges[i], edge_vec)

        # laplace loss
        if i == 0:
            laplace_loss, move_loss = compute_laplace_loss(init_mesh.vertices, coords[i], init_mesh.lap_idx[i], is_first = True)
            laplace_loss = 0.1 * laplace_loss
            
        else:
            laplace_loss, move_loss = compute_laplace_loss(coords[i + 2], coords[i], init_mesh.lap_idx[i], is_first = False)

        total_loss += point_loss + edge_loss * 300 + norm_loss + laplace_loss + move_loss
      
    return total_loss
```

**Comment by ultmaster** [comment]
Sorry that I can't provide the detailed check of your code. But at a first glance, it looks fine to me.

A stupid yet useful debugging method is to convert the checkpoint in the original repo and load it into your code. If it outputs the exact same loss as the original code does, basically it's correct.

There are few tricky things when aligning with tensorflow, e.g., the details when it comes to `Conv2D` stride. I'm sure you will find something useful in our code if you are trying to do this.

If everything is working fine, it might be some problem related to your optimization part. Check your data loading, optimizer parameters and, last but not least, train sufficiently.

These all seem to be standard NN debugging techniques to me, so I'm not sure whether this is helpful.

**Comment by garylidd** [comment]
Thanks for your patience, It's very helpful, I'll try it ~

**Comment by zshyang** [comment]
Hi,
Do you know how to download the dataset and unzip it from the link below?
https://drive.google.com/open?id=131dH36qXCabym1JjSmEpSQZg4dmZVQid

**Comment by zshyang** [comment]
Hi,
Do you know how to use their provided YAML files and pretrained weights downloaded from here to do the evaluation?
To my understanding, I should use the downloaded meta folder. Then use the provide pretrained weights.
The command I use is 
```
python entrypoint_eval.py --name xxx --options experiments/default/tensorflow.yml --checkpoint datasets/data/pretrained/vgg16-p2m.pth 
```
But it seems that it only generated some meaningless meshes. And the score is really low.
I am not sure what I did wrong.
![Screenshot from 2020-09-22 02-08-41](https://user-images.githubusercontent.com/57812785/93863819-de0eac00-fc78-11ea-91b5-0235f5b4487b.png)
